lib/worker.js

```diff
@@ -0,0 +1,211 @@
+'use strict';
+
+if (!process.features.experimental_workers) {
+  throw new Error('Experimental workers are disabled');
+}
+
+const util = require('util');
+const assert = require('assert');
+const EventEmitter = require('events');
+const WorkerContextBinding = process.binding('WorkerContext');
+const JSONStringify = function(value) {
+  if (value === undefined) value = null;
+  return JSON.stringify(value);
+};
+const JSONParse = JSON.parse;
+const EMPTY_ARRAY = [];
+
+const builtinErrorTypes = new Map([
+  Error, SyntaxError, RangeError, URIError, TypeError, EvalError, ReferenceError
+].map(function(Type) {
+  return [Type.name, Type];
+}));
+
+var Worker = WorkerContextBinding.JsConstructor;
+util.inherits(Worker, EventEmitter);
+
+Worker.prototype._init = function(entryModulePath, options) {
+  if (typeof entryModulePath !== 'string')
+    throw new TypeError('entryModulePath must be a string');
+  EventEmitter.call(this);
+  options = Object(options);
+  var keepAlive = options.keepAlive === undefined ? true : !!options.keepAlive;
+  this._workerContext =
+      new WorkerContextBinding.WorkerContext(entryModulePath,
+                                             {keepAlive: keepAlive});
+  this._installEvents();
+};
+
+Worker.prototype._installEvents = function() {
+  const workerObject = this;
+  const workerContext = this._workerContext;
+
+  const onerror = function(payload) {
+    var ErrorConstructor = builtinErrorTypes.get(payload.builtinType);
+    if (typeof ErrorConstructor !== 'function')
+      ErrorConstructor = Error;
+    var error = new ErrorConstructor(payload.message);
+    error.stack = payload.stack;
+    util._extend(error, payload.additionalProperties);
+    workerObject.emit('error', error);
+  };
+
+  workerContext._onexit = function(exitCode) {
+    workerObject._workerContext = null;
+    workerObject.emit('exit', exitCode);
+  };
+
+  workerContext._onmessage = function(payload, messageType) {
+    payload = JSONParse(payload);
+    switch (messageType) {
+      case WorkerContextBinding.USER:
+        return workerObject.emit('message', payload);
+      case WorkerContextBinding.INTERNAL:
+        assert.fail('unreachable');
+      case WorkerContextBinding.EXCEPTION:
+        return onerror(payload);
+      default:
+        assert.fail('unreachable');
+    }
+  };
+};
+
+Worker.prototype._checkAlive = function() {
+  if (!this._workerContext)
+    throw new RangeError('this worker has been terminated');
+};
+
+Worker.prototype._doPostMessage = function(payload, transferList, type) {
+  this._checkAlive();
+
+  if (transferList == null)
+    transferList = EMPTY_ARRAY;
+  else if (!Array.isArray(transferList))
+    throw new Error('transferList must be an array');
+
+  this._workerContext.postMessage(JSONStringify(payload), transferList, type);
+};
+
+Worker.prototype.postMessage = function(payload, callback) {
+  this._doPostMessage(payload, null, WorkerContextBinding.USER);
```

Comment:
1. When transferList is implemented, it becomes the second parameter for this function (the `callback` is not supposed to be there I'll remove it)
2. There used to be `_postInternalMessage` as well but I removed it and didn't refactor :)